### ISSUE TITLE
`FrameworkBuilder` can't return the built `Framework` without starting it

### ISSUE THREAD

**Issue by s0lst1ce** [kangalioo/poise#23]
**What**
The current builder for the framework doesn't provide a method to get the framework without running it.
I believe a `finish()` (or another name like `framework()`) method would be useful.

**Why**
The example from the README proposes (more or less) this code:
```rs
#[tokio::main]
async fn main() {
    poise::Framework::build()
        .token(std::env::var("DISCORD_BOT_TOKEN").unwrap())
        .user_data_setup(move |_ctx, _ready, _framework| {
            Box::pin(async move {
                Ok(())
            })
        })
        .options(poise::FrameworkOptions {
            // configure framework here
            prefix_options: PrefixFrameworkOptions {
                prefix: Some(":t".into()),
                ..Default::default()
            },
            ..Default::default()
        })
        .run()
        .await
        .unwrap();
}
```
This works fine but provides no control over the `Framework` instance which can't even be accessed within `main`. This becomes an issue when you want to have some external control on the state of the framework (running or not). For example if you want to gracefully shutdown the shards when quitting by reacting to an `INTERRUPT` signal. For the intent of making this example clearer the following is code I'd like to be able to write:
```rs
#[tokio::main]
async fn main() {
    let framework = poise::build()
    .token(std::env::var("DISCORD_BOT_TOKEN").unwrap())
    .user_data_setup(move |_ctx, _ready, _framework| {
        Box::pin(async move {
            Ok(())
        })
    })
    .options(poise::FrameworkOptions {
        // configure framework here
        prefix_options: PrefixFrameworkOptions {
            prefix: Some(":t".into()),
            ..Default::default()
        },
        ..Default::default()
    })
    .finish(); // <-- the method I'd like

    tokio::spawn(async move {
        tokio::signal::ctrl_c()
            .await
            .expect("Could not register ctrl+c handler");
        framework.shard_manager().lock().await.shutdown_all().await; // uses framework
    });

    if let Err(why) = framework.start().await { // uses framework
        error!("Client error: {:?}", why);
    }
}
```

Hopefully this clarifies the case for such a method. If I missed something and this can already be done please point me to it. Otherwise I can try to whip a PR for this.

**Comment by kangalioo** [comment]
It's a bit hard because the Framework start() method receives a ClientBuilder, which you have to create and add the password to first.

Possible solutions:
- just deal with that; add a finish() method to FrameworkBuilder and supply the password again in framework start
- make finish() return both the Framework and the prepared ClientBuilder as a tuple probably
- make Framework::start() take Arc<Self>
- make Framework::start() _return_ Arc<Self> (probably doesn't solve your issue though)

2nd solution seems best to me; what do you think?

**Comment by s0lst1ce** [comment]
I think the second solution is the best as well. Maybe make a `PreparedFramework(Framework, Client)`?

**Comment by kangalioo** [comment]
> Maybe make a PreparedFramework(Framework, Client)?

Overkill in my opinion. The only benefit of that function over a plain tuple is that you could make a function on it that calls `framework.start(client)` for you, which is kinda useless

> Should I make a PR then, or have you already made the changes locally?

I'll make the changes locally

---

While hacking on this, I discovered that ClientBuilder's lifetime is actually [a scam](https://github.com/serenity-rs/serenity/pull/1566), and with it gone, the original problem I mentioned is gone. So I went in a small rabbit hole, trying to store ClientBuilder inside Framework, only to discover it doesn't implement Sync because of some obscure API call detail. Then I tried to make it store Client instead, which would require returning `Arc<Framework>` from the new method instead of `Framework`, which is a bit iffy

So, after going on this big tangent I arrived right back where we stopped, because returning a tuple `(Framework, ClientBuilder)` is probably just the best option

**Comment by kangalioo** [comment]
Will this work for you? dddd56d ~~[nice commit name "dddd" :p]~~

**Comment by s0lst1ce** [comment]
Looks good to me.

**Comment by kangalioo** [comment]
Merged to master by the way, feel free to try it

I actually suspect this design won't work after all, because Framework::start takes self by value, which won't work if some other thread needs access to it too

**Comment by kangalioo** [comment]
After thinking again and finally getting a clear mental model of the problem, I implemented a new solution with which your original code would work with slight adjustment (613c2caa95e3173747c3311b3d08f9ee4d3f5276):
```rust
#[tokio::main]
async fn main() {
    let framework = poise::Framework::build()
        /* ... */
        .build().await.unwrap();

    let shard_manager = framework.shard_manager();
    tokio::spawn(async move {
        tokio::signal::ctrl_c()
            .await
            .expect("Could not register ctrl+c handler");
        shard_manager.lock().await.shutdown_all().await;
    });

    if let Err(why) = framework.start().await {
        error!("Client error: {:?}", why);
    }
}
```

<details>
<summary>API design concerncs, in case you're not tired of it by now</summary>

The `.build().await.unwrap()` is a bit ugly...

There's an alternative solution which doesn't need the `.await.unwrap()`. However, that solution doesn't permit accessing the shard manager until after the framework is started. Plus, it requires you to store the client builder (`let (framework, client_builder) = poise::Framework::build()`) to pass it into `framework.start(client_builder)`.

```rust
#[tokio::main]
async fn main() {
    let (framework, client_builder) = poise::Framework::build()
        /* ... */
        .build();

    let framework2 = framework.clone();
    tokio::spawn(async move {
        tokio::signal::ctrl_c()
            .await
            .expect("Could not register ctrl+c handler");
        framework2.shard_manager().lock().await.shutdown_all().await;
    });

    if let Err(why) = framework.start(client_builder).await {
        error!("Client error: {:?}", why);
    }
}
```

So, both solutions have pretty aspects and ugly aspects. Not sure which trade-off is better, maybe you have a preference

</details>

**Comment by s0lst1ce** [comment]
To be frank I hadn't thought about multi-threading.

As for the `build().await.unwrap()` I don't think that's much of an issue. It's not like it'd be a commonly used API anyways.

As for the other option I personally don't know any other use for the shard_manager. Or if it even has a meaning or use before the framework has started? As for the `(framrwork, client_builder)` tuple that's not a big issue. Although I still thing a newtype over it would make sense if it's only used for this.

So really not much of an opinion I guess. Afterall I don't grasp the tradeoffs of each implementation.

**Comment by kangalioo** [comment]
I think the currently implemented API (613c2ca) can be kept:

> As for the build().await.unwrap() I don't think that's much of an issue. It's not like it'd be a commonly used API anyways.

I agree

---

I believe this issue is solved then?

**Comment by s0lst1ce** [comment]
I believe it is yes.
